scipy/sparse/_dia.py

```diff
@@ -176,30 +176,38 @@ def sum(self, axis=None, dtype=None, out=None):
     sum.__doc__ = _spbase.sum.__doc__
 
     def _add_sparse(self, other):
+        # If other is not DIA format, let them handle us instead.
+        if not isinstance(other, _dia_base):
+            return other._add_sparse(self)
+
+        # Fast path for exact equality of the sparsity structure.
+        if np.array_equal(self.offsets, other.offsets):
+            return self._with_data(self.data + other.data)
+
+        # Find the union of the offsets (which will be sorted and unique).
+        new_offsets = np.union1d(self.offsets, other.offsets)
+        self_idx = np.searchsorted(new_offsets, self.offsets)
+        other_idx = np.searchsorted(new_offsets, other.offsets)
+
+        self_d = self.data.shape[1]
+        other_d = other.data.shape[1]
+        # Fast path for a sparsity structure where the final offsets are a
+        # permutation of the existing offsets and the diagonal lengths match.
+        if self_d == other_d and len(new_offsets) == len(self.offsets):
+            new_data = self.data[_invert_index(self_idx)]
+            new_data[other_idx, :] += other.data
```

Comment:
This seems to speed up the case when `self` has diagonals that are a subset of `other`'s diagonals.
And the if logic here maybe could avoid checking `self_d == other_d` twice -- but this seems good. 
```suggestion
        if self_d == other_d and len(new_offsets) == len(self.offsets):
            new_data = self.data[_invert_index(self_idx)]
            new_data[other_idx, :] += other.data
        elif self_d == other_d and len(new_offsets) == len(other.offsets):
            new_data = other.data[_invert_index(other_idx)]
            new_data[self_idx, :] += self.data
```

Reply:
Thanks for catching this, you're absolutely right!